### ISSUE TITLE
[Question] When doing multi-GPU training what does the batch rate represent?

### ISSUE THREAD

**Issue by vedantroy** [mosaicml/composer#1437]
I was training on 1xA100. I noticed that when I switched to using 8xA100, the progress bar indicated instead of doing around 2s/ba, I was doing 3s/ba, which would be slower! I checked that all 8 GPUs were being fully utilized.

My assumption is that the progress indicator has not been updated, so it is not actually 3s/ba, rather it is 3s/8 batches. But, I would like to verify.

Also, I assume the gradient update step is still going to be the same (i.e composer is not changing it underneath the hood).

**Comment by abhi-mosaic** [comment]
Couple parts here:

* When training with Composer we update the weights according to the batch size defined by your dataloaders, no changes under the hood.

* the ProgressBarLogger, and any other loggers such as WandB, will log step == batch == global batch across all devices. Basically when 1 gradient step happens across N devices, we count that as just 1 step. 

* So if you have been increasing the global batch size (e.g. on 1xA100 using a global_batch_size=32 and on 8xA100 using a global_batch_size=256) then I think it's possible the time per step will be a bit higher due to sync overhead.

* for the best experience with distributed training, I would recommend attaching the `SpeedMonitor` which will log `throughput/samples_per_sec` which is a lot more interpretable as you vary # devices and batch size. Could you try that and let us know if `samples_per_sec` increases as exepected from 1xA100 to 8xA100?

**Comment by vedantroy** [comment]
> Couple parts here:
> 
> * When training with Composer we update the weights according to the batch size defined by your dataloaders, no changes under the hood.
> * the ProgressBarLogger, and any other loggers such as WandB, will log step == batch == global batch across all devices. Basically when 1 weight update step happens across N devices, we count that as just 1 step.
> * If you have been increasing the global batch size (e.g. on 1xA100 using a global_batch_size=32 and on 8xA100 using a global_batch_size=256) then it's likely that the time per step will be a bit higher due to gradient sync overhead.
>   
>   * This can happen often with custom entrypoints if you are constructing a dataloader per-device with a line like `DataLoader(..., batch_size=32)` and if that device batch size is not being adjusted with N devices. We're actively working on a way to make this easier with the Trainer.
> * for the best experience with distributed training, I would recommend attaching the `SpeedMonitor` which will log `throughput/samples_per_sec` which is a lot more interpretable as you vary # devices and batch size. Could you try that and let us know if `samples_per_sec` increases as expected from 1xA100 to 8xA100?

@abhi-mosaic, so if I have 8 GPUs and I want to update my gradients every 4096 samples. What should I set my per-device dataloader batch size to be? Should it just be 4096? Or it should it be 4096/8 = 512.

**Comment by abhi-mosaic** [comment]
> @abhi-mosaic, so if I have 8 GPUs and I want to update my gradients every 4096 samples. What should I set my per-device dataloader batch size to be? Should it just be 4096? Or it should it be 4096/8 = 512. My guess is, it is the 2nd.

Yup! You'll want each dataloader to emit 512 samples each.
